### PR TITLE
Only filter ignored paths from module specifier generation if there exists a better option

### LINKED ISSUE #42785: Assertion violation: undefined `relativeSpecifiers` on `Object.getModuleSpecifiers`
From @dlannoye:

```
dependencies_path\typescript@4.2.1-rc\node_modules\typescript\lib\tsc.js:91576
                    ts.Debug.checkDefined(relativeSpecifiers);
                             ^
Error: Debug Failure.
    at Object.getModuleSpecifiers (dependencies_path\typescript@4.2.1-rc\node_modules\typescript\lib\tsc.js:91576:30)
    at getSpecifierForModuleSymbol (dependencies_path\typescript@4.2.1-rc\node_modules\typescript\lib\tsc.js:39998:62)
    at dependencies_path\typescript@4.2.1-rc\node_modules\typescript\lib\tsc.js:39874:39
    at Array.map (<anonymous>)
    at getSymbolChain (dependencies_path\typescript@4.2.1-rc\node_modules\typescript\lib\tsc.js:39872:58)
    at lookupSymbolChainWorker (dependencies_path\typescript@4.2.1-rc\node_modules\typescript\lib\tsc.js:39858:51)
    at lookupSymbolChain (dependencies_path\typescript@4.2.1-rc\node_modules\typescript\lib\tsc.js:39852:24)
    at symbolToTypeNode (dependencies_path\typescript@4.2.1-rc\node_modules\typescript\lib\tsc.js:40005:29)
    at typeToTypeNodeHelper (dependencies_path\typescript@4.2.1-rc\node_modules\typescript\lib\tsc.js:39150:28)
    at mapToTypeNodes (dependencies_path\typescript@4.2.1-rc\node_modules\typescript\lib\tsc.js:39671:40)

```

### LINKED ISSUE #43011: Debug Failure
# Bug Report

Upgrade in `v4.2.2` (also an issue in `v4.2.1-rc`, tsc fail to compile with an error ` Debug Failure.`
### 🕗 Version & Regression Information

* This is a crash
* This changed between versions v4.2.1-beta and v4.2.2-rc
* This is the behaviour in every version I tried, and I reviewed the FAQ for entries about "Debug Failure."

### 💻 Code

[Maskbook](https://github.com/DimensionDev/Maskbook/tree/feature/upgrade-typescript)

### 🙁 Actual behavior

```
Error: Debug Failure.
    at Object.getModuleSpecifiers (/Maskbook/node_modules/.pnpm/typescript@4.2.2/node_modules/typescript/lib/tsc.js:91587:30)
    at getSpecifierForModuleSymbol (/Maskbook/node_modules/.pnpm/typescript@4.2.2/node_modules/typescript/lib/tsc.js:39998:62)
    at createExpressionFromSymbolChain (/Maskbook/node_modules/.pnpm/typescript@4.2.2/node_modules/typescript/lib/tsc.js:40172:63)
    at symbolToExpression (/Maskbook/node_modules/.pnpm/typescript@4.2.2/node_modules/typescript/lib/tsc.js:40159:24)
    at /Maskbook/node_modules/.pnpm/typescript@4.2.2/node_modules/typescript/lib/tsc.js:38966:106
    at withContext (/Maskbook/node_modules/.pnpm/typescript@4.2.2/node_modules/typescript/lib/tsc.js:39005:37)
    at symbolToExpression (/Maskbook/node_modules/.pnpm/typescript@4.2.2/node_modules/typescript/lib/tsc.js:38966:28)
    at symbolToStringWorker (/Maskbook/node_modules/.pnpm/typescript@4.2.2/node_modules/typescript/lib/tsc.js:38886:30)
    at Object.usingSingleLineStringWriter (/Maskbook/node_modules/.pnpm/typescript@4.2.2/node_modules/typescript/lib/tsc.js:10151:13)
    at symbolToString (/Maskbook/node_modules/.pnpm/typescript@4.2.2/node_modules/typescript/lib/tsc.js:38884:73)


```

### 🙂 Expected behavior

It should compile.

### PR DESCRIPTION
**Pull request by andrewbranch**
Fixes #42785 
Fixes #43011

Sometimes the resolver can’t come up with a module specifier that doesn’t have an “ignored path” (like `.pnpm` or `.prisma`) in it. In those cases, we need to use the undesirable ignored path. In @dlannoye’s repro, the `.pnpm` path never actually manifests in a declaration file—the resolver separately finds an accessible re-export and uses that, but there was nevertheless an assertion that we could come up with _something_ for the declaring module itself. The test case included is based off of @flybayer’s repro. This case _does_ result in a type serialization that references `node_modules/.prisma` directly, which isn’t great, but that would have happened in 4.1 and earlier as well. (The real example from @flybayer has `noEmit` enabled, so in fact neither real-world case actually exhibits bad behavior with this PR. And I think for the Prisma case, it’s not strictly harmful if declaration files reference `".prisma"` anyway, since that folder will have to get created during build/install for `"@prisma/client"` to work.)

### CODE REVIEW

File: src/compiler/moduleSpecifiers.ts [hunk 1]
```diff
@@ -282,10 +282,12 @@ namespace ts.moduleSpecifiers {
         const redirects = host.redirectTargetsMap.get(importedPath) || emptyArray;
         const importedFileNames = [...(referenceRedirect ? [referenceRedirect] : emptyArray), importedFileName, ...redirects];
         const targets = importedFileNames.map(f => getNormalizedAbsolutePath(f, cwd));
+        let filterIgnoredPaths = some(targets, not(containsIgnoredPath));
```
Comment (DanielRosenwasser) [inline src/compiler/moduleSpecifiers.ts:285]: Alternatively

```suggestion
        let filterIgnoredPaths = !every(targets, containsIgnoredPath);
```

but yours reads better.

I might have called it

`shouldFilterIgnoredPaths` though.

**Review by weswigham — APPROVED** [review]
Is the assertion that triggered the crash still correct to have? Is this function no longer able to return `undefined` in the cases that feed into that assertion?

**Comment by andrewbranch** [comment]
Yes. Fixes #43011.

**Comment by andrewbranch** [comment]
> Is the assertion that triggered the crash still correct to have? Is this function no longer able to return `undefined` in the cases that feed into that assertion?

Yeah, the assertion is correct, and the function I changed is now guaranteed to call its callback at least once. (If we skip the filtering, `targets` always contains at least `importedFileName` and will run on that, unless better symlinks are found.)

**Comment by andrewbranch** [comment]
@typescript-bot cherry-pick this to release-4.2

**Comment by andrewbranch** [comment]
@typescript-bot cherry-pick this to release-4.2

**Review by sheetalkamat — APPROVED** [review]

**Pull request merged**